zenith_utils/src/pq_proto.rs

```diff
@@ -0,0 +1,559 @@
+// Postgres protocol messages serialization-deserialization
+
+use anyhow::{anyhow, bail};
+use byteorder::{ReadBytesExt, WriteBytesExt, BE};
+use bytes::{Buf, BufMut, Bytes, BytesMut};
+// use postgres_ffi::xlog_utils::TimestampTz;
+use std::collections::HashMap;
+use std::io;
+use std::io::{Read, Write};
+use std::str;
+
+pub type Oid = u32;
+pub type SystemId = u64;
+
+#[derive(Debug)]
+pub enum FeMessage {
+    StartupMessage(FeStartupMessage),
+    Query(FeQueryMessage), // Simple query
+    Parse(FeParseMessage), // Extended query protocol
+    Describe(FeDescribeMessage),
+    Bind(FeBindMessage),
+    Execute(FeExecuteMessage),
+    Close(FeCloseMessage),
+    Sync,
+    Terminate,
+    CopyData(Bytes),
+}
+
+#[derive(Debug)]
+pub struct FeStartupMessage {
+    pub version: u32,
+    pub kind: StartupRequestCode,
+    // optional params arriving in startup packet
+    pub params: HashMap<String, String>,
+}
+
+#[derive(Debug)]
+pub enum StartupRequestCode {
+    Cancel,
+    NegotiateSsl,
+    NegotiateGss,
+    Normal,
+}
+
+#[derive(Debug)]
+pub struct FeQueryMessage {
+    pub body: Bytes,
+}
+
+// We only support the simple case of Parse on unnamed prepared statement and
+// no params
+#[derive(Debug)]
+pub struct FeParseMessage {
+    pub query_string: Bytes,
+}
+
+#[derive(Debug)]
+pub struct FeDescribeMessage {
+    pub kind: u8, // 'S' to describe a prepared statement; or 'P' to describe a portal.
+                  // we only support unnamed prepared stmt or portal
+}
+
+// we only support unnamed prepared stmt and portal
+#[derive(Debug)]
+pub struct FeBindMessage {}
```

Comment:
Should we make this an empty struct?

Reply:
You mean unit struct `struct FeBindMessage;`? I would leave this as is, we might support this stuff in the future, and here we don't seem to be cared by unit/empty struct difference.